Support fan speed levels

blejdfist opened the pull request:
## Description:


**Related issue (if applicable):** 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1056
**Pull request in [aioesphomeapi](https://github.com/esphome/aioesphomeapi):** https://github.com/esphome/aioesphomeapi/pull/27
**Pull request in [homeassistant core](https://github.com/home-assistant/core/):** https://github.com/home-assistant/core/pull/46712

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

blejdfist committed:
Add fan speed percentage support to the API

blejdfist committed:
Add float fan speed percentage

blejdfist committed:
Add percentage support to automation and configuration

blejdfist committed:
Update Tuya fan

blejdfist committed:
Fix pylint warning

0x0a11c0de commented:
It looks like this PR is still hardcoding 3 fan speeds in both the speed fan and the tuya fan. The whole point of the change to speed percentage in HA was to support more than 3 fan speeds. My tuya fans have 4 speeds, for example.

0x0a11c0de commented:
Since the code has to convert from percentage to speed settings, it needs to know the number of valid speeds. To get rid of the current hardcoded speed range of 0 - 2, i.e. three speeds, a config variable needs to be added to set the number of speeds. If that needs to be added, why do percentages at all? It would be simpler to just add a `speed_count` config variable and change the FanSpeed enum to an int. HA already does the conversion from percentage to integers, all it needs is the `speed_count`. Doing it this way would require a minimum of code changes on both ends to support all fan speeds.

blejdfist commented:
> 
> 
> Since the code has to convert from percentage to speed settings, it needs to know the number of valid speeds. To get rid of the current hardcoded speed range of 0 - 2, i.e. three speeds, a config variable needs to be added to set the number of speeds. If that needs to be added, why do percentages at all? It would be simpler to just add a `speed_count` config variable and change the FanSpeed enum to an int. [HA already does the conversion from percentage to integers](https://github.com/home-assistant/core/blob/10dae253e5c10b0f2feecf899d6d0fae426b1043/homeassistant/components/esphome/fan.py#L73), all [it needs is the `speed_count`](https://github.com/home-assistant/core/blob/10dae253e5c10b0f2feecf899d6d0fae426b1043/homeassistant/components/esphome/fan.py#L123). Doing it this way would require a minimum of code changes on both ends to support all fan speeds.

Absolutely, this is the plan. The change to Tuya fan was only to preserve compatibility, as well as provide the method to map the percentage range onto a discrete range of values.

blejdfist committed:
Update API to use speed levels instead of percentage

blejdfist committed:
Use speed levels

blejdfist committed:
Merge branch 'dev' into support_fan_percentage

blejdfist committed:
Fix type warnings

blejdfist committed:
MQTT component now converts between speed levels and enums

blejdfist committed:
Webserver now supports speed_level

blejdfist committed:
Update prometheus

blejdfist committed:
Remove low/medium/high settings from speed fan

blejdfist committed:
Merge branch 'dev' into support_fan_percentage

# Conflicts:
#	esphome/components/fan/__init__.py
#	esphome/components/speed/fan/__init__.py

blejdfist committed:
Remove unused enum

glmnet commented:
This needs updates on docs

glmnet commented:
Also a template fan is needed now, there was an old PR about this. Now speed fan is hard coded to 100 levels and there were many pseudo hacks which converted low mid high to a realy seet output, a custom template fan which accepts arbitrary speed level and a lambda action to act is needed now

blejdfist commented:
> Also a template fan is needed now, there was an old PR about this. Now speed fan is hard coded to 100 levels and there were many pseudo hacks which converted low mid high to a realy seet output, a custom template fan which accepts arbitrary speed level and a lambda action to act is needed now.
> 
> This can be in a separate PR, this one is ok as is, just needs docs

I've updated the docs, hopefully I didn't miss something.
Do you happen to know the number of that old PR?

glmnet commented:
I might be confused. Seems like it doesn't exists.

blejdfist commented:
> I might be confused. Seems like it doesn't exists.

Alright, we could still add one. Might be a useful workaround if some people want to replicate the old behavior for some reason.

glmnet commented:
Another option would be to make the speed_count a parameter in speed fan

glmnet commented:
What about speed names. Are we dropping that? It does make sense for low med and high only, otherwise people naturally uses numbers

blejdfist commented:
> Another option would be to make the speed_count a parameter in speed fan
> 
> However that doesn't fit nice on an analog output. But it's how is working now
> 
> A custom output is used which translates the analog value to a select on which relay to turn on

I was planning on doing that, but not just for Speed Fan, also for Tuya for example since many people have Tuya fans with more than 3 speeds. For speed fan, if you set the speed count to, say, 3. Then the three speeds would be 0.33, 0.66 and 1.0.
Default would be 100 for speed fan, and 3 for Tuya.

What do you think?


> What about speed names. Are we dropping that? It does make sense for low med and high only, otherwise people naturally uses numbers

Yes, I removed that. It is my understanding that the only reason we had low/medium/high was because HA had that. HA has now dropped that in favour of only using a percentage for speed, and a speed count for any fans that have discrete levels (still uses percentage though).

There is support for preset modes in HA though, but these are not intended for speeds but rather things like different modes. My fan for example has a mode that changes speed every few minutes or so to simulate "natural wind". 

Or did you have another idea in mind?

glmnet commented:
Ok then. I'll say we add configurable speed count on speed fan, this give more flexibility. 

Previous implementation allowed custom output values for low mid high which will not be supported anymore so template fan is the way to go IMO

blejdfist commented:
> Ok then. I'll say we add configurable speed count on speed fan, this give more flexibility.
> 
> Previous implementation allowed custom output values for low mid high which will not be supported anymore so template fan is the way to go IMO

Fair enough, should be quick :)

blejdfist committed:
Configurable speed levels for speed fan

blejdfist committed:
Remove unused import

glmnet commented on esphome/components/speed/fan/speed_fan.h:
```diff
@@ -24,6 +25,7 @@ class SpeedFan : public Component {
   output::BinaryOutput *oscillating_{nullptr};
   output::BinaryOutput *direction_{nullptr};
   bool next_update_{true};
+  int speed_levels_{100};
```
We don't put defaults on initialized members

blejdfist replied:
Fixed.

Regarding "speed_levels", no not really. I had a different name to not be confused with the old "speed".
But I guess it would make sense to name them `speed_count` and `speed`.

glmnet replied:
That will be more consistent with hass and it better gives a sense of discrete levels support

blejdfist replied:
Btw, note that in the protocol it still has to be something different from "speed" as some applications might rely on that field.

glmnet reviewed (commented):
I just realized you are using `speed_levels` instead of `speed_count` is there a reason for that?

blejdfist committed:
Rename speed_level->speed and speed_levels->speed_count

blejdfist committed:
Rename supported_speed_levels -> supported_speed_count in API and FanTraits

Field id stays the same in the protocol, so the change is not breaking for aioesphome.

glmnet commented:
Awesome, I think is ready now, thanks
